Framing serial

korken89 opened the pull request:
~Work in progress PR for framing serial with DMA.~

PR for framing serial with character match using DMA.

korken89 commented:
@MabezDev / @MathiasKoch I have implemented the framing based DMA / UART transfers that utilize character match.

Some things I want to do:

* Move from slice operation to raw operations, the compiler cannot remove all panicing branches based on slice access.
* Use the input array as if it was `MaybeUninit` so uninitialized data can be used. No good reason to set data to a value when the DMA later will overwrite it.

Feedback is welcome!

Here is an example program:

```rust
#![no_main]
#![no_std]

// use panic_persist as _;
//use panic_halt as _;
use cortex_m::asm;
use cortex_m_semihosting::hprintln;
pub use firmware as _;
use hal::{
    dma,
    prelude::*,
    rcc::{ClockSecuritySystem, CrystalBypass, MsiFreq},
    serial::{self, Config, Serial},
};
use heapless::{
    pool,
    pool::singleton::{Box, Pool},
};
use panic_semihosting as _;
use rtfm::app;
use stm32l4xx_hal as hal;

pool!(
    A: [u8; 8]
);

#[app(device = stm32l4xx_hal::stm32, peripherals = true)]
const APP: () = {
    struct Resources {
        rx: serial::Rx<hal::stm32::USART2>,
        frame_reader: dma::FrameReader<Box<A>, dma::dma1::C6>,
    }

    #[init]
    fn init(cx: init::Context) -> init::LateResources {
        static mut MEMORY: [u8; 1024] = [0; 1024];
        A::grow(MEMORY);

        let dp = cx.device;
        let mut flash = dp.FLASH.constrain();
        let mut rcc = dp.RCC.constrain();
        let mut pwr = dp.PWR.constrain(&mut rcc.apb1r1);
        let mut gpioa = dp.GPIOA.split(&mut rcc.ahb2);

        let clocks = rcc
            .cfgr
            .lse(CrystalBypass::Disable, ClockSecuritySystem::Disable)
            .msi(MsiFreq::RANGE4M)
            .sysclk(80.mhz())
            .freeze(&mut flash.acr, &mut pwr);

        let tx2 = gpioa.pa2.into_af7(&mut gpioa.moder, &mut gpioa.afrl);
        let rx2 = gpioa.pa3.into_af7(&mut gpioa.moder, &mut gpioa.afrl);

        // TRY using a different USART peripheral here
        let mut serial = Serial::usart2(
            dp.USART2,
            (tx2, rx2),
            Config::default()
                .baudrate(115_200.bps())
                .character_match(b'a'),
            clocks,
            &mut rcc.apb1r1,
        );
        serial.listen(serial::Event::CharacterMatch);
        let (mut tx, serial_rx) = serial.split();

        let channels = dp.DMA1.split(&mut rcc.ahb1);
        let mut dma_ch6 = channels.6;
        dma_ch6.listen(dma::Event::TransferComplete);

        let dma_buf = A::alloc().unwrap().init([0; 8]);
        let fr = serial_rx.frame_read(dma_ch6, dma_buf);

        let sent = b'X';
        tx.write(sent).ok();

        init::LateResources {
            rx: serial_rx,
            frame_reader: fr,
        }
    }

    #[idle]
    fn idle(_: idle::Context) -> ! {
        hprintln!("IDLE").ok();

        loop {
            asm::nop();
        }
    }

    #[task(binds = USART2, resources = [rx, frame_reader], priority = 3)]
    fn serial_isr(cx: serial_isr::Context) {
        // Check for character match
        if cx.resources.rx.is_character_match(true) {
            let dma_buf = A::alloc().unwrap().init([0; 8]);
            let (buf, len) = cx.resources.frame_reader.character_match_interrupt(dma_buf);
            hprintln!("CM: {}", unsafe {
                core::str::from_utf8_unchecked(&buf[0..len])
            })
            .ok();

            // NB: Memory is return to the pool here when it goes out of scope
        }
    }

    #[task(binds = DMA1_CH6, resources = [rx, frame_reader], priority = 3)]
    fn serial_dma_isr(cx: serial_dma_isr::Context) {
        let dma_buf = A::alloc().unwrap().init([0; 8]);

        let (buf, len) = cx
            .resources
            .frame_reader
            .transfer_complete_interrupt(dma_buf);
        hprintln!("DMA: {}", unsafe {
            core::str::from_utf8_unchecked(&buf[0..len])
        })
        .ok();

        // NB: Memory is return to the pool here when it goes out of scope
    }
};
```

korken89 commented:
The new version addressing my thoughts have been pushed, here is a new example program.
The enticing feature here is that this is without copies, a buffer is ever copied while received.

Now to fix sending of frames. :)

```
#![no_main]
#![no_std]

use cortex_m::asm;
use cortex_m_rt::{self};
use cortex_m_semihosting::hprintln;
use hal::{
    dma::{self, consts, FrameReader, SerialDMAFrame},
    prelude::*,
    rcc::{ClockSecuritySystem, CrystalBypass, MsiFreq},
    serial::{self, Config, Serial},
};
use heapless::{
    pool,
    pool::singleton::{Box, Pool},
};
use panic_semihosting as _;
use rtfm::app;
use stm32l4xx_hal as hal;

pool!(
    A: SerialDMAFrame<consts::U8>
);

#[app(device = stm32l4xx_hal::stm32, peripherals = true)]
const APP: () = {
    struct Resources {
        rx: serial::Rx<hal::stm32::USART2>,
        frame_reader: FrameReader<Box<A>, dma::dma1::C6, consts::U8>,
        nothing: (),
    }

    #[init]
    fn init(cx: init::Context) -> init::LateResources {
        static mut MEMORY: [u8; 1024] = [0; 1024];

        // increase the capacity of the pool
        A::grow(MEMORY);

        let dp = cx.device;

        let mut flash = dp.FLASH.constrain();
        let mut rcc = dp.RCC.constrain();
        let mut pwr = dp.PWR.constrain(&mut rcc.apb1r1);
        let mut gpioa = dp.GPIOA.split(&mut rcc.ahb2);

        let clocks = rcc
            .cfgr
            .lse(CrystalBypass::Disable, ClockSecuritySystem::Disable)
            .msi(MsiFreq::RANGE4M)
            .sysclk(80.mhz())
            .freeze(&mut flash.acr, &mut pwr);

        let tx2 = gpioa.pa2.into_af7(&mut gpioa.moder, &mut gpioa.afrl);
        let rx2 = gpioa.pa3.into_af7(&mut gpioa.moder, &mut gpioa.afrl);

        // TRY using a different USART peripheral here
        let mut serial = Serial::usart2(
            dp.USART2,
            (tx2, rx2),
            Config::default()
                .baudrate(115_200.bps())
                .character_match(b'a'),
            clocks,
            &mut rcc.apb1r1,
        );
        serial.listen(serial::Event::CharacterMatch);
        let (mut tx, serial_rx) = serial.split();

        let channels = dp.DMA1.split(&mut rcc.ahb1);
        let mut dma_ch6 = channels.6;
        dma_ch6.listen(dma::Event::TransferComplete);

        let dma_buf = A::alloc().unwrap().init(SerialDMAFrame::new());
        let fr = serial_rx.frame_read(dma_ch6, dma_buf);

        let sent = b'X';

        tx.write(sent).ok();

        init::LateResources {
            rx: serial_rx,
            frame_reader: fr,
            nothing: (),
        }
    }

    #[idle]
    fn idle(_: idle::Context) -> ! {
        hprintln!("IDLE").ok();

        loop {
            asm::nop();
        }
    }

    #[task(binds = USART2, resources = [rx, frame_reader], priority = 3)]
    fn serial_isr(cx: serial_isr::Context) {
        hprintln!("UISR").ok();

        // Check for character match
        if cx.resources.rx.is_character_match(true) {
            let dma_buf = A::alloc().unwrap().init(SerialDMAFrame::new());
            let buf = cx.resources.frame_reader.character_match_interrupt(dma_buf);
            hprintln!("CM: {}", unsafe {
                core::str::from_utf8_unchecked(buf.read())
            })
            .ok();
        }
    }

    #[task(binds = DMA1_CH6, resources = [rx, frame_reader], priority = 3)]
    fn serial_dma_isr(cx: serial_dma_isr::Context) {
        hprintln!("TX").ok();

        let dma_buf = A::alloc().unwrap().init(SerialDMAFrame::new());

        let buf = cx
            .resources
            .frame_reader
            .transfer_complete_interrupt(dma_buf);
        hprintln!("DMA: {}", unsafe {
            core::str::from_utf8_unchecked(buf.read())
        })
        .ok();
    }
};
```

korken89 commented:
I have now pushed so sending and receiving is working and I added an example for this using RTFM.

I think this is ready for review now.

MathiasKoch commented:
Whats with the `nothing: ()`?

Other than that, i can't wait to test this out!

korken89 commented:
Ah right, for got to remove that. Fixed now.
It's for keeping a stable interface while testing :)

korken89 commented:
I added receiver timeout support and made some documentation improvements

thalesfragoso commented on src/serial.rs:
```diff
@@ -675,24 +906,19 @@ macro_rules! hal {
                 ) -> CircBuffer<B, $rx_chan>
                 where
-                    B: StableDeref<Target = [H; 2]> + DerefMut,
+                    B: StableDeref<Target = [H; 2]> + DerefMut + 'static,
```
Maybe change the `StableDeref` to just `Deref` ?

MabezDev replied:
Yeah I don't think `StableDeref` is required anymore; I can't actually remeber why I pulled it in to behonest. This change can probably wait though, I wouldn't want to block this PR over it.

thalesfragoso replied:
`StableDeref` kinda helps, but it isn't enough to get rid of the `mem::forget` problem. The way it is now (without the `'static`) is unsound. I guess it's a breaking change to put `'static`, but we have to do it.
We probably don't need `StableDeref` now that we got `'static`, so we might as well change it to `Deref`.

thalesfragoso commented on src/serial.rs:
```diff
@@ -711,28 +937,154 @@ macro_rules! hal {
+                    // TODO can we weaken this compiler barrier?
+                    // NOTE(compiler_fence) operations on `buffer` should not be reordered after
+                    // the next statement, which starts the DMA transfer
+                    atomic::compiler_fence(Ordering::SeqCst);
```
```suggestion
                    atomic::compiler_fence(Ordering::Release);
```

thalesfragoso commented on src/serial.rs:
```diff
@@ -711,28 +937,154 @@ macro_rules! hal {
+
+                    // NOTE(compiler_fence) operations on `buffer` should not be reordered after
+                    // the next statement, which starts the DMA transfer
+                    atomic::compiler_fence(Ordering::SeqCst);
```
```suggestion
                    atomic::compiler_fence(Ordering::Release);
```

MabezDev commented:
This is looking really great! Not seeing any real issues, other than the ones kindly reported, thank you @thalesfragoso!

korken89 commented on src/dma.rs:
```diff
@@ -197,42 +414,228 @@ macro_rules! dma {
+                            }
+
+                            self.channel.stop();
+
```
`compiler_fence(Ordering::SeqCst)` did not change the size of the code, so used that.

korken89 commented:
Thanks for the comment @thalesfragoso, I have updated accordingly!

korken89 commented:
The changes to the compiler fenced did not produce a different binary in my case, but better with weaker ones if it is OK. :)

korken89 commented:
@thalesfragoso Interesting approach! I have to give a deeper look to that.

I did a final commit to this PR.
1) Changed name to `DMAFrame`, the frame has no connection to the `serial` so the name was misleading.
2) Added a `write` and `commit` method, some serialization frameworks wants an `&mut [u8]` which now is possible.

thalesfragoso commented on src/dma.rs:
```diff
@@ -145,6 +145,26 @@ where
+    ///
+    /// Note that this function internally first zeros the node's buffer.
+    pub fn write(&mut self) -> &mut [u8] {
+        // Initialize memory with a safe value
+        self.buf = unsafe { core::mem::zeroed() };
+        self.len = self.max_len() as u16;
+
```
Maybe don't zero all the buffer ? Something like this
```rust
for elem in self.buf.iter_mut().skip(self.len) {
    *elem = MaybeUninit::zeroed();
}
```

thalesfragoso commented:
> @thalesfragoso Interesting approach! I have to give a deeper look to that.

I changed a lot of things in that initial PR, so don't look much at the module docs, I still have to change that.

MathiasKoch commented:
Works beautifully here, with both the frame reader and receiver timeout for receiving AT commands! Thanks!

korken89 commented:
Thanks for the review @thalesfragoso, I have updated the code!
I think this is ready for merge now. :)

@MathiasKoch Awesome, glad to hear that you are already getting use from it!

korken89 commented:
@MabezDev Whenever you feel ready, this should now be ready for merge.

MabezDev commented:
@korken89 I was just writing up a comment asking if this was good to go :joy:. This looks really great, thanks for the contribution!
